packages/ui-keyring/src/index.ts

```diff
@@ -4,83 +4,429 @@
+        );
+        const [, hexAddr] = key.split(':');
+
+        addresses.add(address, json);
+
+        if (hexAddr.substr(0, 2) !== '0x') {
+          store.remove(key);
+          store.set(addressKey(hexAddr), json);
+        }
+      }
+    });
+
+    this.initOptions();
+  }
+
+  restoreAccount (json: KeyringPair$Json, password: string): KeyringPair {
+    const pair = createPair(
+      {
+        publicKey: decodeAddress(json.address),
+        secretKey: new Uint8Array()
+      },
+      json.meta,
+      hexToU8a(json.encoded)
+    );
+
+    // unlock, save account and then lock (locking cleans secretKey, so needs to be last)
+    pair.decodePkcs8(password);
+    this.addAccountPair(pair, password);
+    pair.lock();
+
+    return pair;
+  }
+
+  saveAccount (pair: KeyringPair, password?: string): void {
+    const { accounts, keyring } = this.state;
+    const json = pair.toJson(password);
+
+    if (!json.meta.whenCreated) {
+      json.meta.whenCreated = Date.now();
+    }
+
+    keyring.addFromJson(json);
+    accounts.add(json.address, json);
+  }
+
+  saveAccountMeta (pair: KeyringPair, meta: KeyringPair$Meta): void {
+    const { accounts } = this.state;
+    const address = pair.address();
+    const json = store.get(accountKey(address));
+
+    pair.setMeta(meta);
+    json.meta = pair.getMeta();
+
+    accounts.add(json.address, json);
+  }
+
+  saveAddress (address: string, meta: KeyringPair$Meta): void {
+    const { addresses } = this.state;
+    const available = addresses.subject.getValue();
+
+    const json = (available[address] && available[address].json) || {
+      address,
+      meta: {
+        isRecent: void 0,
+        whenCreated: Date.now()
+      }
+    };
+
+    Object.keys(meta).forEach((key) => {
+      json.meta[key] = meta[key];
+    });
+
+    delete json.meta.isRecent;
+
+    addresses.add(address, json);
+  }
+
+  saveRecent (address: string): SingleAddress {
+    const { addresses } = this.state;
+    const available = addresses.subject.getValue();
+
+    if (!available[address]) {
+      const json = {
+        address,
+        meta: {
+          isRecent: true,
+          whenCreated: Date.now()
+        }
+      };
+
+      addresses.add(address, (json as KeyringJson));
+    }
+
+    return addresses.subject.getValue()[address];
+  }
+
+  setDevMode (isDevelopment: boolean): void {
+    observableDevelopment.set(isDevelopment);
+  }
+
+  private addAccounts (options: KeyringOptions): void {
+    const { accounts } = this.state;
+    const accountsAvailable = accounts.subject.getValue();
+
+    Object
+      .keys(accountsAvailable)
+      .map((address) =>
+        accountsAvailable[address]
+      )
+      .forEach(({ json: { meta: { isTesting = false } }, option }: SingleAddress) => {
+        if (!isTesting) {
+          options.account.push(option);
+        } else {
+          options.testing.push(option);
+        }
+      });
+  }
+}
+
+Keyring.counter = 0;
+
+export {
+  Keyring
```

Comment:
Rather don't export at all. (Only the instance, not the class - no hoops needed above then)

Reply:
I needed to export the class for the unit tests https://github.com/polkadot-js/apps/pull/387/files/f45d32fb7cb33d5bf14676cbb57072d22edabdee#diff-ef3c29ea5508f636947d9a7544f8bf7dR5

Reply:
But exporting the class for the unit tests exposes it to someone creating multiple instances of it

Reply:
Actually, since `initOptions` is moving elsewhere I won't need to test that in Keyring class anymore